---
title: "figure_5"
author: "Tabea M. Soelter"
date: "2024-01-28"
output: html_document
---
**Plotting Manuscript Figure 5**

__Goal__: This is a script for generating figure 5 of my manuscript. We are also saving supplementary table 4.

__Reproducibility__:
* GitHub: lasseignelab/230418_TS_AgingCCC
* Docker: tsoelter/rstudio_aging_ccc
    * Version: 1.0.2
* HPC: Yes
    * Resources: long partition (150hrs), 1 GPU, 6 CPUs, 65GB per CPU

__Data__: 
*MultiNicheNet object*
* Name: multinichenet_output.rds
* Location: data/ccc/
*Filtered DESeq2 output*
* Name: filtered_dea_targets.rds
* Location: results/final_outputs/04_dea/
*AD Gene List*
* Name: ad_genes_mm.rds
* Location: doc/

__Analysis Plan__:
* Load necessary packages
* Load data
  * Load AD gene lists
  * MultiNicheNet object
  * DESeq2 outputs
* Filter all input data
* Plot individual panels:
  * A: Alluvial plot of LRs where targets are AD genes
  * B: Heatmap of targets that are AD genes showing their differential gex
  * C: Alluvial plot of LRs where either the ligand or the receptor is an AD gene
* Compile figure
* Save figure

# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# set seed
set.seed(42)

# load packages
suppressPackageStartupMessages({
  library(tidyverse)
  library(ggalluvial)
  library(ComplexHeatmap)
  library(circlize)
  library(RColorBrewer)
  library(patchwork)
  library(cowplot)
  library(ggpubr)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCin3xTgAD.R"))
```

# Colors
```{r}
colors <- c(
  `Oligodendrocytes` = "royalblue3",
  `Microglia` = "orchid3",
  `OPCs` = "mediumpurple1",
  `Astrocytes` = "cornflowerblue",
  `Excitatory.Neurons` = "slateblue3",
  `Inhibitory.Neurons` = "darkslategray3",
  `Pericytes` = "deepskyblue3",
  `Endothelial cells` = "darkolivegreen4",
  `Meningeal cells` = "seagreen4",
  `Ependymal cells` = "lightcoral",
  `Fibroblasts` = "olivedrab3",
  `RGCs` = "violetred2"
)
```

# Load Data
* Filtered DESeq2 output
* MultiNicheNet output
* AD risk gene list
```{r}
# Target gene expression info from DESeq2
all_gex <- readRDS(here(
  "results",
  "final_outputs",
  "04_dea",
  "filtered_dea_targets.rds"
))

# MultiNicheNet object
output <- readRDS(here("data", "ccc", "multinichenet_output.rds"))

# AD gene list
ad_genes_mm <- readRDS(here("doc", "ad_genes_mm.rds"))
```

# Filter MultiNicheNet output
```{r}
# number of targets
top_n_target <- 250

# contrast table
contrast_tbl <- tibble(
  contrast = c("X6mAD-X6mWT", "X6mWT-X6mAD", "X12mAD-X12mWT", "X12mWT-X12mAD"),
  group = c("X6mAD", "X6mWT", "X12mAD", "X12mWT")
)

# set receiver and sender cell types of interest
receiver_oi <- c("Excitatory.Neurons", "Inhibitory.Neurons")
sender_oi <- c("Astrocytes", "Microglia", "OPCs", "Oligodendrocytes")

# vector of groups of interest
group_oi <- c("X6mAD", "X6mWT", "X12mAD", "X12mWT")

# Filter output using above criteria and add time point column
output_filt <- filter_nichenet(output)
```

# Filter DEA info by AD-associated genes
```{r}
targets_ad <- all_gex %>%
  filter(gene %in% ad_genes_mm)

# filter by significance
sig_ad_targets <- targets_ad %>%
  filter(padj < 0.05)
```

# Prepare df for plotting
```{r}
# Filter MultiNicheNet object by AD genes
ad_lrt <- output_filt %>%
  select(sender, receiver, ligand, receptor, target, group) %>%
  filter(target %in% sig_ad_targets$gene) %>%
  unique()

# Factorize columns of interest (for plotting purposes)
vec <- c("ligand", "receptor", "target")

ad_lrt[vec] <- lapply(ad_lrt[vec], function(x) factor(x, levels = unique(x)))

# Reorder columns of interest by the sender to aid with legibility
ad_lrt$ligand <- reorder(
  ad_lrt$ligand,
  desc(ad_lrt$sender)
)
ad_lrt$receptor <- reorder(
  ad_lrt$receptor,
  desc(ad_lrt$sender)
)
```

# Figure 5A
* Alluvial plot of LRs that have an AD-associated target gene
```{r}
# Filter for LRs only
lr_all <- ad_lrt %>%
  select(sender, ligand, receptor) %>%
  unique()

# Plot
alluvial_lr_all <- ggplot(lr_all,
  aes(axis1 = ligand, axis2 = receptor),
  label = stratum
) +
  scale_x_discrete(limits = c("Ligand", "Receptor")) +
  xlab("Type") +
  geom_alluvium() +
  stat_alluvium(lode.guidance = "frontback") +
  geom_stratum(aes(fill = sender)) +
  geom_label(
    stat = "stratum",
    fill = "white",
    aes(label = after_stat(stratum)),
    size = 3,
    fontface = "bold"
  ) +
  geom_flow(aes(fill = sender), stat = "alluvium", color = "black") +
  scale_fill_manual(values = colors) +
  scale_x_discrete(limits = c("Ligand", "Receptor")) +
  theme_void() +
  theme(
    axis.text.x = element_text(
      vjust = 2.5,
      face = "bold",
      size = 12,
      hjust = 0.5
    ),
    legend.position = c(0.9, 0.5),
    legend.margin = margin(0, 0, 0, 0.5, "cm"),
    legend.text = element_text(face = "bold", size = 10),
    legend.title = element_text(face = "bold", size = 12),
    plot.background = element_rect(fill = "white", size = 0),
    plot.margin = grid::unit(c(-0.4, 0.3, 0, -2.5), "cm"),
    axis.ticks = element_blank(),
    axis.ticks.length = unit(0, "pt"),
    axis.text.y = element_blank(),
    axis.title.y = element_blank(),
    panel.spacing = element_blank(),
    axis.title.x = element_blank()
  ) +
  guides(fill = guide_legend(title = "Sender"))

alluvial_lr_all

# save plot
ggsave(
  here(
    "results",
    "final_outputs",
    "03_ccc",
    "alluvial_lr_all.png"
  ),
  alluvial_lr_all,
  dpi = 300,
  width = 8,
  height = 8,
  bg = "#ffffff"
)
```

# Figure 5B
* Heatmap of targets that are also AD genes
```{r}
# Prepare input df (pivot to create gene by celltype/timepoint matrix)
target_df <- targets_ad %>%
  select(gene, log2FoldChange, padj, receiver, timepoint) %>%
  pivot_wider(
    id_cols = gene,
    names_from = c(receiver, timepoint),
    values_from = c(log2FoldChange, padj)
  ) %>%
  mutate(gene_sig = case_when(
    padj_Excitatory.Neurons_12m < 0.05 ~ paste0(gene, "^"),
    padj_Inhibitory.Neurons_12m < 0.05 ~ paste0(gene, "*"),
    padj_Inhibitory.Neurons_6m < 0.05 ~ paste0(gene, "*"),
    TRUE ~ gene
  )) %>%
  select(-c(
    padj_Excitatory.Neurons_12m,
    padj_Inhibitory.Neurons_12m,
    padj_Inhibitory.Neurons_6m,
    gene
  )) %>%
  column_to_rownames(var = "gene_sig")

targets_ad$receiver <- gsub("\\.", " ", targets_ad$receiver)



target_df <- targets_ad %>%
  select(gene, log2FoldChange, receiver, timepoint) %>%
  pivot_wider(
    id_cols = gene,
    names_from = c(receiver, timepoint),
    values_from = log2FoldChange
  ) %>%
  column_to_rownames(var = "gene")

target_df <- target_df[, col_order]

target_df2 <- targets_ad %>%
  select(gene, padj, receiver, timepoint) %>%
  pivot_wider(
    id_cols = gene,
    names_from = c(receiver, timepoint),
    values_from = padj
  ) %>%
  column_to_rownames(var = "gene")

target_df2[is.na(target_df2)] <- "NS"

target_df2 <- target_df2[, col_order]

mat2 <- as.matrix(target_df2)

# Replace NAs with 0
target_df[is.na(target_df)] <- 0

# Prepare heatmap annotation
meta <- as.data.frame(colnames(target_df))
rownames(meta) <- meta$`colnames(target_df)`

#meta$receiver <- sub("^[^_]*_([^_]*)_.+", "\\1", colnames(target_df))
#meta$timepoint <- sub(".*_", "", colnames(target_df))

#meta <- meta %>%
#  select(receiver, timepoint)

meta$cell_type <- sub("_.*", "", colnames(target_df))
meta$timepoint <- sub(".*_", "", colnames(target_df))
#meta$cell_type <- gsub("\\.", " ", meta$cell_type)

meta <- meta %>%
  select(cell_type, timepoint)

# Set annotation colors
anno_cols <- list(
  "cell_type" = c(
    "Excitatory Neurons" = "slateblue3",
    "Inhibitory Neurons" = "darkslategray3"
  ),
  "timepoint" = c(
    "6m" = "#6A3D9A",
    "12m" = "#CAB2D6"
  )
)

# Generate annotation
anno <- HeatmapAnnotation(
  df = meta,
  show_annotation_name = FALSE,
  col = anno_cols,
  annotation_legend_param = list(
    cell_type = list(
      title = "Receiver",
      direction = "horizontal",
      labels = c("Excitatory Neurons", "Inhibitory Neurons"),
      labels_gp = gpar(fontface = "bold", fontsize = 10),
      title_gp = gpar(fontface = "bold", fontsize = 12)
    ),
    timepoint = list(
      title = "Time Point",
      labels_gp = gpar(fontface = "bold", fontsize = 10),
      title_gp = gpar(fontface = "bold", fontsize = 12)
    )
  )
)

# Convert df to matrix for plotting
mat <- as.matrix(target_df)

# Set heatmap colors
#cols <- colorRamp2(c(-0.5, 0, 0.5), rev(brewer.pal(n = 3, name = "BrBG")))
cols <- colorRamp2(c(-1, -0.5, 0, 0.5, 1), c("darkblue", "lightblue", "grey95", "tomato", "darkred"))

# Plot
set.seed(42)
gex_heatmap <- Heatmap(mat,
  col = cols,
  top_annotation = anno,
  show_column_names = FALSE,
  show_row_dend = TRUE,
  show_column_dend = FALSE,
  cluster_columns = TRUE,
  cluster_rows = TRUE,
  row_title = "AD-associated Target Genes",
  row_title_gp = gpar(fontface = "bold", fontsize = 14),
  row_names_gp = gpar(fontface = "bold", fontsize = 10),
  heatmap_legend_param = list(
    direction = "horizontal",
    title = "Differential Expression",
    labels_gp = gpar(fontface = "bold", fontsize = 10),
    title_gp = gpar(fontface = "bold", fontsize = 12)
  )
)

draw(gex_heatmap,
  heatmap_legend_side = "right",
  annotation_legend_side = "right",
  merge_legend = TRUE
)

# grab plot and legends for combining plot
heatmap <- grid.grabExpr(
  draw(gex_heatmap,
    heatmap_legend_side = "right",
    annotation_legend_side = "right",
    merge_legend = TRUE
  )
)

# save plot
ggsave(
  here(
    "results",
    "final_outputs",
    "03_ccc",
    "heatmap_ad_targets.png"
  ),
  heatmap,
  dpi = 300,
  width = 6,
  height = 6,
  bg = "#ffffff"
)
```

# Figure 5C
* Filter for LRs that are AD genes
* So either the ligand OR the receptor has to be an AD gene
```{r}
# Include only LRs where either the ligand or the receptor is an AD gene
ad_lr <- output_filt %>%
  select(sender, receiver, ligand, receptor, target, group) %>%
  filter(ligand %in% ad_genes_mm | receptor %in% ad_genes_mm) %>%
  unique()

ad_lr_filt <- ad_lr %>%
  select(sender, ligand, receptor) %>%
  unique()

# Plot
alluvial_lr_ad <- ggplot(ad_lr_filt,
  aes(axis1 = ligand, axis2 = receptor),
  label = stratum
) +
  scale_x_discrete(limits = c("Ligand", "Receptor")) +
  xlab("Type") +
  geom_alluvium() +
  stat_alluvium(lode.guidance = "frontback") +
  geom_stratum(aes(fill = sender)) +
  geom_label(
    stat = "stratum",
    fill = "white",
    aes(label = after_stat(stratum)),
    size = 3,
    fontface = "bold"
  ) +
  geom_flow(aes(fill = sender), stat = "alluvium", color = "black") +
  scale_fill_manual(values = colors) +
  scale_x_discrete(limits = c("Ligand", "Receptor")) +
  theme_void() +
  theme(
    axis.text.x = element_text(vjust = 2, face = "bold", size = 12),
    legend.position = c(0.9, 0.5),
    legend.margin = margin(0, 0, 0, -0.2, "cm"),
    legend.text = element_text(face = "bold", size = 10),
    legend.title = element_text(face = "bold", size = 12),
    plot.background = element_rect(fill = "white", size = 0),
    plot.margin = grid::unit(c(0.5, 0, 0, -3), "cm"),
    axis.ticks = element_blank(),
    axis.ticks.length = unit(0, "pt"),
    axis.text.y = element_blank(),
    axis.title.y = element_blank(),
    panel.spacing = element_blank(),
    axis.title.x = element_blank()
  ) +
  guides(fill = guide_legend(title = "Sender"))

alluvial_lr_ad

# save plot
ggsave(
  here(
    "results",
    "final_outputs",
    "03_ccc",
    "alluvial_lr_ad.png"
  ),
  alluvial_lr_ad,
  dpi = 300,
  width = 6,
  height = 6,
  bg = "#ffffff"
)
```

# Combine figure 5
```{r}
left <- ggarrange(alluvial_lr_all,
  labels = "A"
)

right <- ggarrange(heatmap,
  alluvial_lr_ad,
  nrow = 2,
  labels = c("B", "C")
)

# Combine left and right sides, align horizontally and vertically
fig5 <- ggarrange(left,
  right,
  align = "hv",
  ncol = 2
)

## testing
fig5 <- ggarrange(alluvial_lr_all,
  heatmaps,
  align = "hv",
  ncol = 2,
  labels = c("A", "B")
)

```

# Save figure 5
```{r}
ggsave(
  here(
    "results",
    "final_outputs",
    "01_figures",
    "figure5_v2.png"
  ),
  fig5,
  dpi = 300,
  width = 13,
  height = 8,
  bg = "#ffffff"
)
```

# Session info
```{r}
sessionInfo() # see output below
```
R version 4.2.3 (2023-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lintr_3.0.2           styler_1.9.1          here_1.0.1            ggpubr_0.6.0         
 [5] cowplot_1.1.1         patchwork_1.1.2       RColorBrewer_1.1-3    circlize_0.4.15      
 [9] ComplexHeatmap_2.14.0 ggalluvial_0.12.5     gprofiler2_0.2.1      lubridate_1.9.2      
[13] forcats_1.0.0         stringr_1.5.0         dplyr_1.1.1           purrr_1.0.1          
[17] readr_2.1.4           tidyr_1.3.0           tibble_3.2.1          ggplot2_3.4.2        
[21] tidyverse_2.0.0      

loaded via a namespace (and not attached):
 [1] colorspace_2.1-0    ggsignif_0.6.4      rjson_0.2.21        rprojroot_2.0.3    
 [5] GlobalOptions_0.1.2 clue_0.3-64         rstudioapi_0.14     farver_2.1.1       
 [9] remotes_2.4.2       bit64_4.0.5         fansi_1.0.4         xml2_1.3.3         
[13] codetools_0.2-19    R.methodsS3_1.8.2   doParallel_1.0.17   knitr_1.42         
[17] jsonlite_1.8.4      Cairo_1.6-0         broom_1.0.4         cluster_2.1.4      
[21] png_0.1-8           R.oo_1.25.0         compiler_4.2.3      httr_1.4.5         
[25] backports_1.4.1     fastmap_1.1.1       lazyeval_0.2.2      cli_3.6.1          
[29] htmltools_0.5.5     tools_4.2.3         gtable_0.3.3        glue_1.6.2         
[33] carData_3.0-5       vctrs_0.6.2         iterators_1.0.14    xfun_0.38          
[37] ps_1.7.5            timechange_0.2.0    lifecycle_1.0.3     WriteXLS_6.4.0     
[41] rstatix_0.7.2       scales_1.2.1        vroom_1.6.1         ragg_1.2.5         
[45] hms_1.1.3           rex_1.2.1           parallel_4.2.3      yaml_2.3.7         
[49] stringi_1.7.12      S4Vectors_0.36.2    desc_1.4.2          foreach_1.5.2      
[53] BiocGenerics_0.44.0 cyclocomp_1.1.0     shape_1.4.6         rlang_1.1.0        
[57] pkgconfig_2.0.3     systemfonts_1.0.4   matrixStats_0.63.0  bitops_1.0-7       
[61] evaluate_0.20       htmlwidgets_1.6.2   labeling_0.4.2      bit_4.0.5          
[65] processx_3.8.1      tidyselect_1.2.0    magrittr_2.0.3      R6_2.5.1           
[69] IRanges_2.32.0      generics_0.1.3      pillar_1.9.0        withr_2.5.0        
[73] abind_1.4-5         RCurl_1.98-1.12     crayon_1.5.2        car_3.1-2          
[77] utf8_1.2.3          plotly_4.10.1       tzdb_0.3.0          rmarkdown_2.21     
[81] GetoptLong_1.0.5    data.table_1.14.8   callr_3.7.3         digest_0.6.31      
[85] R.cache_0.16.0      R.utils_2.12.2      textshaping_0.3.6   stats4_4.2.3       
[89] munsell_0.5.0       viridisLite_0.4.1

# Time tracking
```{r}
fptm <- proc.time() - ptm

# convert to minutes
(fptm[3] / 60)

# Time elapsed: 10 minutes
```

# Reproducibility
```{r}
# styler
style_file(here(
  "src",
  "figures",
  "figure_5.Rmd"
))

# lintr
lint(filename = here(
  "src",
  "figures",
  "figure_5.Rmd"
))
```
