---
title: "figure_S3"
author: "Tabea M. Soelter"
date: "2024-01-28"
output: html_document
---
**Plotting Manuscript Supplementary Figure 3**

__Goal__: This is a script for generating figure S3 of my manuscript, where I plot AD gene expression across all cell types of interest.

__Reproducibility__:
* GitHub: lasseignelab/230418_TS_AgingCCC
* Docker: tsoelter/rstudio_aging_ccc
    * Version: 1.0.2
* HPC: Yes
    * Resources: long partition (150hrs), 1 GPU, 6 CPUs, 65GB per CPU

__Data__: 
*DESeq2 outputs*
* Names:
    1. *celltype*_group_6mAD_vs_6mWT_all_genes.csv
      * For every cell type and time point 
*Converted AD gene list*
* Name: ad_genes_mm.rds
* Location: doc/

__Analysis Plan__:
* Load necessary packages
* Load data
  * AD gene list
  * DESeq2 outputs
* Filter all input data
* Plot Heatmap
  * single panel
* Save figure

# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# set seed
set.seed(42)

# load packages
suppressPackageStartupMessages({
  library(tidyverse)
  library(readr)
  library(ComplexHeatmap)
  library(circlize)
  library(RColorBrewer)
  library(patchwork)
  library(cowplot)
  library(ggpubr)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCin3xTgAD.R"))
```

# Load in AD gene list
```{r}
ad_genes_mm <- readRDS(here("doc", "ad_genes_mm.rds"))
```

# Load DESeq2 outputs & combine
```{r include=FALSE}
astrocytes_6m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "astrocytes_group_6mAD_vs_6mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Astrocytes") %>%
  mutate(time_point = "6m")

astrocytes_12m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "astrocytes_group_12mAD_vs_12mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Astrocytes") %>%
  mutate(time_point = "12m")

excitatory_neurons_6m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "excitatory_neurons_group_6mAD_vs_6mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Excitatory Neurons") %>%
  mutate(time_point = "6m")

excitatory_neurons_12m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "excitatory_neurons_group_12mAD_vs_12mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Excitatory Neurons") %>%
  mutate(time_point = "12m")

inhibitory_neurons_6m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "inhibitory_neurons_group_6mAD_vs_6mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Inhibitory Neurons") %>%
  mutate(time_point = "6m")

inhibitory_neurons_12m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "inhibitory_neurons_group_12mAD_vs_12mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Inhibitory Neurons") %>%
  mutate(time_point = "12m")

microglia_6m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "microglia_group_6mAD_vs_6mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Microglia") %>%
  mutate(time_point = "6m")

microglia_12m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "microglia_group_12mAD_vs_12mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Microglia") %>%
  mutate(time_point = "12m")

oligodendrocytes_6m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "oligodendrocytes_group_6mAD_vs_6mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Oligodendrocytes") %>%
  mutate(time_point = "6m")

oligodendrocytes_12m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "oligodendrocytes_group_12mAD_vs_12mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "Oligodendrocytes") %>%
  mutate(time_point = "12m")

opcs_6m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "opcs_group_6mAD_vs_6mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "OPCs") %>%
  mutate(time_point = "6m")

opcs_12m <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "opcs_group_12mAD_vs_12mWT_all_genes.csv"
)) %>%
  mutate(cell_type = "OPCs") %>%
  mutate(time_point = "12m")

# combine all DESeq2 outputs into 1 df
combined_gex <- rbind(
  astrocytes_6m,
  astrocytes_12m,
  excitatory_neurons_6m,
  excitatory_neurons_12m,
  inhibitory_neurons_6m,
  inhibitory_neurons_12m,
  microglia_6m,
  microglia_12m,
  oligodendrocytes_6m,
  oligodendrocytes_12m,
  opcs_6m,
  opcs_12m
)

# remove individual DEA objects from environment, as they're no longer needed
rm(list = ls(pattern = "6m"))
rm(list = ls(pattern = "12m"))
```

# Filter DESeq2 outputs
```{r}
# Make cell type and time point vectors
cell_type_oi <- as.vector(unique(combined_gex$cell_type))
time_point_oi <- as.vector(unique(combined_gex$time_point))

# Filter DEA output for significant AD gex by cell type and time point
ad_genes_dge <- combined_gex %>%
  filter(cell_type %in% cell_type_oi, time_point %in% time_point_oi) %>%
  filter(gene %in% ad_genes_mm, padj < 0.05)
```

# Plot figure S3
* Large heatmap of all AD genes that are significantly differentially expressed (padj)
* In cell types of interest
```{r}
dea_df <- ad_genes_dge %>%
  select(gene, log2FoldChange, cell_type, time_point) %>%
  pivot_wider(
    id_cols = gene,
    names_from = c(cell_type, time_point),
    values_from = log2FoldChange
  ) %>%
  column_to_rownames(var = "gene")

# Replace NAs with 0
dea_df[is.na(dea_df)] <- 0

# Prepare heatmap annotation
meta <- as.data.frame(colnames(dea_df))
rownames(meta) <- meta$`colnames(dea_df)`

meta$cell_type <- sub("_.*", "", colnames(dea_df))
meta$timepoint <- sub(".*_", "", colnames(dea_df))

meta <- meta %>%
  select(cell_type, timepoint)

# Set annotation colors
anno_cols <- list(
  "cell_type" = c(
    "Oligodendrocytes" = "royalblue3",
    "Microglia" = "orchid3",
    "OPCs" = "mediumpurple1",
    "Astrocytes" = "cornflowerblue",
    "Excitatory Neurons" = "slateblue3",
    "Inhibitory Neurons" = "darkslategray3"
  ),
  "timepoint" = c(
    "6m" = "#6A3D9A",
    "12m" = "#CAB2D6"
  )
)

# Generate annotation
anno <- HeatmapAnnotation(
  df = meta,
  show_annotation_name = FALSE,
  col = anno_cols,
  annotation_legend_param = list(
    cell_type = list(
      title = "Cell Type",
      direction = "horizontal",
      labels_gp = gpar(fontface = "bold", fontsize = 10),
      title_gp = gpar(fontface = "bold", fontsize = 12)
    ),
    timepoint = list(
      title = "Time Point",
      direction = "horizontal",
      labels_gp = gpar(fontface = "bold", fontsize = 10),
      title_gp = gpar(fontface = "bold", fontsize = 12)
    )
  )
)

# Convert df to matrix
mat <- as.matrix(dea_df)

# Set heatmap colors
cols <- colorRamp2(c(-0.5, 0, 0.5), rev(brewer.pal(n = 3, name = "RdBu")))

set.seed(42)
heatmap <- Heatmap(mat,
  col = cols,
  top_annotation = anno,
  show_column_names = FALSE,
  show_row_dend = TRUE,
  show_column_dend = FALSE,
  cluster_columns = FALSE,
  cluster_rows = TRUE,
  row_title = "AD Genes",
  row_title_gp = gpar(fontface = "bold", fontsize = 14),
  row_names_gp = gpar(fontface = "bold", fonstize = 10),
  heatmap_legend_param = list(
    direction = "horizontal",
    title = "Differential Expression",
    labels_gp = gpar(fontface = "bold", fontsize = 10),
    title_gp = gpar(fontface = "bold", fontsize = 12)
  )
)

draw(heatmap,
  heatmap_legend_side = "right",
  annotation_legend_side = "right",
  merge_legend = TRUE
)
```

# Save figure
```{r}
# Grab heatmap with legends to save
heatmap_ad <- grid.grabExpr(
  draw(heatmap,
    heatmap_legend_side = "right",
    annotation_legend_side = "right",
    merge_legend = TRUE
  )
)

# save
ggsave(
  here(
    "results",
    "final_outputs",
    "01_figures",
    "supp_figure3.png"
  ),
  heatmap_ad,
  dpi = 300,
  width = 10,
  height = 10,
  bg = "#ffffff"
)
```

# Session info
```{r}
sessionInfo() # see output below
```
R version 4.2.3 (2023-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lintr_3.0.2           styler_1.9.1          here_1.0.1            ggpubr_0.6.0         
 [5] cowplot_1.1.1         patchwork_1.1.2       RColorBrewer_1.1-3    circlize_0.4.15      
 [9] ComplexHeatmap_2.14.0 lubridate_1.9.2       forcats_1.0.0         stringr_1.5.0        
[13] dplyr_1.1.1           purrr_1.0.1           readr_2.1.4           tidyr_1.3.0          
[17] tibble_3.2.1          ggplot2_3.4.2         tidyverse_2.0.0      

loaded via a namespace (and not attached):
 [1] matrixStats_0.63.0  bit64_4.0.5         doParallel_1.0.17   rprojroot_2.0.3    
 [5] R.cache_0.16.0      tools_4.2.3         backports_1.4.1     utf8_1.2.3         
 [9] R6_2.5.1            BiocGenerics_0.44.0 lazyeval_0.2.2      colorspace_2.1-0   
[13] GetoptLong_1.0.5    withr_2.5.0         tidyselect_1.2.0    processx_3.8.1     
[17] bit_4.0.5           compiler_4.2.3      textshaping_0.3.6   cli_3.6.1          
[21] Cairo_1.6-0         xml2_1.3.3          desc_1.4.2          scales_1.2.1       
[25] callr_3.7.3         systemfonts_1.0.4   digest_0.6.31       rmarkdown_2.21     
[29] R.utils_2.12.2      htmltools_0.5.5     pkgconfig_2.0.3     WriteXLS_6.4.0     
[33] fastmap_1.1.1       rlang_1.1.0         GlobalOptions_0.1.2 rstudioapi_0.14    
[37] shape_1.4.6         generics_0.1.3      vroom_1.6.1         car_3.1-2          
[41] R.oo_1.25.0         magrittr_2.0.3      munsell_0.5.0       S4Vectors_0.36.2   
[45] fansi_1.0.4         abind_1.4-5         lifecycle_1.0.3     R.methodsS3_1.8.2  
[49] yaml_2.3.7          stringi_1.7.12      carData_3.0-5       parallel_4.2.3     
[53] crayon_1.5.2        hms_1.1.3           knitr_1.42          ps_1.7.5           
[57] pillar_1.9.0        rjson_0.2.21        ggsignif_0.6.4      codetools_0.2-19   
[61] stats4_4.2.3        glue_1.6.2          evaluate_0.20       rex_1.2.1          
[65] remotes_2.4.2       png_0.1-8           vctrs_0.6.2         tzdb_0.3.0         
[69] foreach_1.5.2       gtable_0.3.3        clue_0.3-64         xfun_0.38          
[73] broom_1.0.4         cyclocomp_1.1.0     rstatix_0.7.2       ragg_1.2.5         
[77] iterators_1.0.14    IRanges_2.32.0      cluster_2.1.4       timechange_0.2.0  

# Time tracking
```{r}
fptm <- proc.time() - ptm

# convert to minutes
(fptm[3] / 60)

# Time elapsed: 5 minutes
```

# Reproducibility
```{r}
# styler
style_file(here(
  "src",
  "figures",
  "figure_S3.Rmd"
))

# lintr
lint(filename = here(
  "src",
  "figures",
  "figure_S3.Rmd"
))
```
