---
title: "figure_4"
author: "Tabea M. Soelter"
date: "2024-01-28"
output: html_document
---
**Plotting Manuscript Figure 4**

__Goal__: This is a script for generating figure 4 of my manuscript. I am also generating Table S3.

__Reproducibility__:
* GitHub: lasseignelab/230418_TS_AgingCCC
* Docker: tsoelter/rstudio_aging_ccc
    * Version: 1.0.2
* HPC: Yes
    * Resources: long partition (150hrs), 1 GPU, 6 CPUs, 65GB per CPU

__Data__:
*MultiNicheNet object*
* Name: multinichenet_output.rds
* Location: data/ccc/
*DESeq2 outputs*
* Names:
    1. excitatory_neurons_group_6mAD_vs_6mWT_all_genes.csv
    2. inhibitory_neurons_group_6mAD_vs_6mWT_all_genes.csv
    3. excitatory_neurons_group_12mAD_vs_12mWT_all_genes.csv
    4. inhibitory_neurons_group_12mAD_vs_12mWT_all_genes.csv
* Location: results/intermediate_outputs/03_dea/
*FEA gprofiler2 outputs*
* Names:
    1. fea_res_6m.rds
    2. fea_res_12m.rds
* Location: results/intermediate_outputs/04_ccc/

__Analysis Plan__:
* Load necessary packages
* Load data
  * MultiNicheNet object
  * DESeq2 outputs
  * FEA gprofiler2 outputs
* Save Table S3
* Plot individual panels:
  * A: Volcano plot of gex of all targets (across time points and conditions)
  * B: Bubble plot for 6-month FEA
  * C: Bubble plot for 12-month FEA
* Compile figure
* Save figure

# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# set seed
set.seed(42)

# load packages
suppressPackageStartupMessages({
  library(tidyverse)
  library(WriteXLS)
  library(ggrepel)
  library(scales)
  library(patchwork)
  library(cowplot)
  library(ggpubr)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCin3xTgAD.R"))
```

# Load Data
## Load MultiNicheNet object
```{r}
output <- readRDS(here("data", "ccc", "multinichenet_output.rds"))
```

## Load DESeq2 outputs
```{r}
# excitatory neurons
ex_6m_ad_all <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "excitatory_neurons_group_6mAD_vs_6mWT_all_genes.csv"
))

ex_12m_ad_all <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "excitatory_neurons_group_12mAD_vs_12mWT_all_genes.csv"
))

# inhibitory neurons
in_6m_ad_all <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "inhibitory_neurons_group_6mAD_vs_6mWT_all_genes.csv"
))

in_12m_ad_all <- read_csv(here(
  "results",
  "intermediate_outputs",
  "03_dea",
  "inhibitory_neurons_group_12mAD_vs_12mWT_all_genes.csv"
))
```

## Load FEA outputs
```{r}
fea_res_6m <- readRDS(here(
  "results",
  "intermediate_outputs",
  "04_ccc",
  "fea_res_6m.rds"
))

fea_res_12m <- readRDS(here(
  "results",
  "intermediate_outputs",
  "04_ccc",
  "fea_res_12m.rds"
))
```

# Filter MultiNicheNet output
```{r}
# number of targets
top_n_target <- 250

# contrast table
contrast_tbl <- tibble(
  contrast = c("X6mAD-X6mWT", "X6mWT-X6mAD", "X12mAD-X12mWT", "X12mWT-X12mAD"),
  group = c("X6mAD", "X6mWT", "X12mAD", "X12mWT")
)

# set receiver and sender cell types of interest
receiver_oi <- c("Excitatory.Neurons", "Inhibitory.Neurons")
sender_oi <- c("Astrocytes", "Microglia", "OPCs", "Oligodendrocytes")

# vector of groups of interest
group_oi <- c("X6mAD", "X6mWT", "X12mAD", "X12mWT")

# Filter output using above criteria and add time point column
output_filt <- filter_nichenet(output) %>%
  mutate(time_point = ifelse(grepl("X6m", group), "6m", "12m"))
```

# Filter DESeq2 outputs by group-specific targets
* We filtered DESeq2 outputs (which are cell type and group-specific) by the target genes specific to those groupings.
```{r}
# Filter 6-month targets
in_6m_gex <- filter_dea(output_filt,
  in_6m_ad_all,
  receiver_oi = "Inhibitory.Neurons",
  timepoint_oi = "6m"
)

ex_6m_gex <- filter_dea(output_filt,
  ex_6m_ad_all,
  receiver_oi = "Excitatory.Neurons",
  timepoint_oi = "6m"
)

# Filter 12-month targets
in_12m_gex <- filter_dea(output_filt,
  in_12m_ad_all,
  receiver_oi = "Inhibitory.Neurons",
  timepoint_oi = "12m"
)

ex_12m_gex <- filter_dea(output_filt,
  ex_12m_ad_all,
  receiver_oi = "Excitatory.Neurons",
  timepoint_oi = "12m"
)

# Combine group-specific gex dfs into one
all_gex <- rbind(
  ex_6m_gex,
  ex_12m_gex,
  in_6m_gex,
  in_12m_gex
)
```

# Save Table S3
* Save filtered DEA output
```{r}
# save for future analyses
saveRDS(
  all_gex,
  here("results", "final_outputs", "04_dea", "filtered_dea_targets.rds")
)

# save as excel file for manuscript (Supplementary Table 3)
WriteXLS(
  all_gex,
  here("results", "final_outputs", "01_figures", "table_s3.xlsx")
)
```

# Figure 4A
* Volcano plot of all targets GEX
* In order to color the volcano plot differently, we added gex information"
  * NS/DE: Genes that are not differentially expressed (DE), majority are also not significant (NS)
    * log2FoldChange between -0.2 and 0.2, padj > 0.05 for majority
  * NS up: Genes that are DE (upregulated), but NS according to multiple hypothesis correction
    * log2FoldChange > 0.2, padj > 0.05
  * NS down: Genes that are DE (downregulated), but NS according to multiple hypothesis correction
    * log2FoldChange < -0.2, padj > 0.05
  * Sig up/down: Genes that are significantly DE
    * up: log2FoldChange > 0.2, padj < 0.05
    * down: log2FoldChange < -0.2, padj < 0.05
```{r}
# adding expression information to df
all_gex$diffexpressed <- "NS/DE"

all_gex$diffexpressed[all_gex$log2FoldChange > 0.2 &
  all_gex$padj < 0.05] <- "Significantly up"

all_gex$diffexpressed[all_gex$log2FoldChange < -0.2 &
  all_gex$padj < 0.05] <- "Significantly down"

all_gex$diffexpressed[all_gex$log2FoldChange > 0.2 &
  all_gex$padj > 0.05] <- "NS up"

all_gex$diffexpressed[all_gex$log2FoldChange < -0.2 &
  all_gex$padj > 0.05] <- "NS down"


# add labels to significantly DE genes
all_gex$delabel <- NA
all_gex$delabel[all_gex$diffexpressed != "NS/DE"] <-
  all_gex$gene[all_gex$diffexpressed != "NS/DE"]

# plot volcano plot
volcano <- ggplot(
  all_gex,
  aes(
    x = log2FoldChange,
    y = -log10(padj),
    col = diffexpressed,
    label = delabel
  )
) +
  geom_point() +
  theme_bw() +
  geom_text_repel() +
  scale_color_manual(values = c(
    "lightblue",
    "tomato",
    "grey",
    "darkblue",
    "darkred"
  )) +
  geom_vline(
    xintercept = c(-0.2, 0.2),
    col = "black",
    linetype = "dotted"
  ) +
  geom_hline(
    yintercept = -log10(0.05),
    col = "black",
    linetype = "dotted"
  ) +
  labs(
    x = "Differential Expression (log2FoldChange)",
    y = "Adjusted p-value (-log10)",
    color = "Target Gene Expression"
  ) +
  guides(color = guide_legend(override.aes = list(shape = 19, size = 3))) +
  theme(
    axis.text = element_text(face = "bold", color = "black", size = 10),
    axis.title = element_text(face = "bold", color = "black", size = 12),
    legend.title = element_text(face = "bold", size = 12),
    legend.text = element_text(face = "bold", size = 10)
  )

volcano

# save volcano plot
ggsave(
  here(
    "results",
    "final_outputs",
    "04_dea",
    "volcano_all_targets.png"
  ),
  volcano,
  dpi = 400,
  width = 9,
  height = 5,
  bg = "#ffffff"
)
```

# Figure 4B
* Bubbleplot of 6m targets' FEA
```{r}
bubbleplot_6m <- ggplot(
  fea_res_6m,
  aes(
    x = condition,
    y = reorder(term_name, p_value),
    size = recall,
    fill = p_value
  )
) +
  geom_point(alpha = 0.7, shape = 21) +
  scale_size(range = c(2, 10), name = "Recall") +
  scale_fill_distiller(
    palette = "Greens",
    limits = c(0, 0.05),
    breaks = c(0, 0.03, 0.05)
  ) +
  labs(
    x = "Condition",
    y = "6m Functional Enrichment Terms",
    fill = "Adj p-value"
  ) +
  theme_bw() +
  theme(
    axis.text = element_text(face = "bold", color = "black", size = 10),
    text = element_text(face = "bold", size = 12),
    axis.text.x = element_text(angle = 0, hjust = 0.5)
  ) +
  scale_y_discrete(labels = wrap_format(15)) +
  coord_flip()

bubbleplot_6m
```

# Figure 4C
* Bubbleplot of 12m targets' FEA
```{r}
bubbleplot_12m <- ggplot(
  fea_res_12m,
  aes(
    x = condition,
    y = reorder(term_name, p_value),
    size = recall,
    fill = p_value
  )
) +
  geom_point(alpha = 0.7, shape = 21) +
  scale_size(range = c(2, 10), name = "Recall") +
  scale_fill_distiller(
    palette = "Greens",
    limits = c(0, 0.05),
    breaks = c(0, 0.03, 0.05)
  ) +
  labs(
    x = "Condition",
    y = "12m Functional Enrichment Terms",
    fill = "Adj p-value"
  ) +
  scale_y_discrete(labels = wrap_format(65)) +
  theme_bw() +
  theme(
    axis.text = element_text(face = "bold", color = "black", size = 10),
    text = element_text(face = "bold", size = 12)
  )

bubbleplot_12m
```

# Compile Figure 4
* Compiling figure by the left (volcano and 6m FEA) and right (12m FEA) plots
```{r}
left <- ggarrange(volcano,
  bubbleplot_6m,
  nrow = 2,
  labels = c("A", "B"),
  align = "v",
  heights = c(1.5, 1)
)

right <- ggarrange(bubbleplot_12m,
  labels = "C"
)

# compile final figure
fig4 <- ggarrange(left,
  right,
  ncol = 2
)

# plot figure
fig4
```

# Save final figure
```{r}
ggsave(
  here(
    "results",
    "final_outputs",
    "01_figures",
    "figure4.png"
  ),
  figv2,
  dpi = 400,
  width = 16,
  height = 9,
  bg = "#ffffff"
)
```

# Session info
```{r}
sessionInfo() # see output below
```
R version 4.2.3 (2023-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lintr_3.0.2     styler_1.9.1    here_1.0.1      ggpubr_0.6.0    cowplot_1.1.1  
 [6] patchwork_1.1.2 scales_1.2.1    ggrepel_0.9.3   WriteXLS_6.4.0  lubridate_1.9.2
[11] forcats_1.0.0   stringr_1.5.0   dplyr_1.1.1     purrr_1.0.1     readr_2.1.4    
[16] tidyr_1.3.0     tibble_3.2.1    ggplot2_3.4.2   tidyverse_2.0.0

loaded via a namespace (and not attached):
  [1] colorspace_2.1-0            ggsignif_0.6.4              rprojroot_2.0.3            
  [4] scuttle_1.8.4               XVector_0.38.0              GenomicRanges_1.50.2       
  [7] rstudioapi_0.14             farver_2.1.1                remotes_2.4.2              
 [10] bit64_4.0.5                 fansi_1.0.4                 xml2_1.3.3                 
 [13] codetools_0.2-19            R.methodsS3_1.8.2           sparseMatrixStats_1.10.0   
 [16] knitr_1.42                  broom_1.0.4                 R.oo_1.25.0                
 [19] HDF5Array_1.26.0            compiler_4.2.3              dqrng_0.3.0                
 [22] backports_1.4.1             Matrix_1.5-4                fastmap_1.1.1              
 [25] lazyeval_0.2.2              limma_3.54.2                cli_3.6.1                  
 [28] htmltools_0.5.5             tools_4.2.3                 gtable_0.3.3               
 [31] glue_1.6.2                  GenomeInfoDbData_1.2.9      Rcpp_1.0.10                
 [34] carData_3.0-5               Biobase_2.58.0              vctrs_0.6.2                
 [37] rhdf5filters_1.10.1         DelayedMatrixStats_1.20.0   xfun_0.38                  
 [40] ps_1.7.5                    beachmat_2.14.2             timechange_0.2.0           
 [43] lifecycle_1.0.3             rstatix_0.7.2               edgeR_3.40.2               
 [46] zlibbioc_1.44.0             vroom_1.6.1                 ragg_1.2.5                 
 [49] hms_1.1.3                   MatrixGenerics_1.10.0       rex_1.2.1                  
 [52] parallel_4.2.3              SummarizedExperiment_1.28.0 rhdf5_2.42.1               
 [55] RColorBrewer_1.1-3          yaml_2.3.7                  SingleCellExperiment_1.20.1
 [58] stringi_1.7.12              S4Vectors_0.36.2            desc_1.4.2                 
 [61] BiocGenerics_0.44.0         cyclocomp_1.1.0             BiocParallel_1.32.6        
 [64] GenomeInfoDb_1.34.9         systemfonts_1.0.4           rlang_1.1.0                
 [67] pkgconfig_2.0.3             matrixStats_0.63.0          bitops_1.0-7               
 [70] evaluate_0.20               lattice_0.21-8              Rhdf5lib_1.20.0            
 [73] labeling_0.4.2              bit_4.0.5                   processx_3.8.1             
 [76] tidyselect_1.2.0            plyr_1.8.8                  magrittr_2.0.3             
 [79] R6_2.5.1                    IRanges_2.32.0              generics_0.1.3             
 [82] DelayedArray_0.24.0         pillar_1.9.0                withr_2.5.0                
 [85] abind_1.4-5                 RCurl_1.98-1.12             crayon_1.5.2               
 [88] car_3.1-2                   DropletUtils_1.18.1         utf8_1.2.3                 
 [91] tzdb_0.3.0                  rmarkdown_2.21              locfit_1.5-9.7             
 [94] grid_4.2.3                  callr_3.7.3                 digest_0.6.31              
 [97] R.cache_0.16.0              textshaping_0.3.6           R.utils_2.12.2             
[100] stats4_4.2.3                munsell_0.5.0 

# Time tracking
```{r}
fptm <- proc.time() - ptm

# convert to minutes
(fptm[3] / 60)

# Time elapsed: 5 minutes
```

# Reproducibility
```{r eval=FALSE, include=FALSE}
# styler
style_file(here(
  "src",
  "figures",
  "figure_4.Rmd"
))

# lintr
lint(filename = here(
  "src",
  "figures",
  "figure_4.Rmd"
))
```
