---
title: "04_grn"
author: "Tabea M. Soelter"
date: "2024-01-26"
output: html_document
---
**Determining signaling mediators using gene regulatory networks from MultiNicheNet**

__Goal__: To identify potential signaling mediators (receptor -> target gene) by determining direct neighbors of receptors.
  
__Reproducibility__: 
* GitHub: lasseignelab/230418_TS_AgingCCC
* Docker: tsoelter/rstudio_aging_ccc
    * Version: 1.0.2
* HPC: Yes
  * Resources: long partition (150hrs), 6 CPUs, 65GB per CPU

__Data__:
*MultiNicheNet object*
* Name: multinichenet_output.rds
* Location: data/ccc/
*AD Gene Lists*
* Names:
    1. HP_ALZHEIMER_DISEASE.v2023.2.Hs.tsv
    2. KEGG_ALZHEIMERS_DISEASE.v2023.2.Hs.tsv
    3. ad_gene_list.csv
* Location: doc/
 
__Analysis Plan__:
* Load necessary packages 
* Load data
* Filter MultiNicheNet output
* Grab AD-associated LRTs
* Extract signaling GRNs from MultiNicheNet
  * Plot and save
* Create igraph objects for each LRT
* Calculate network properties
  * Receptor neighbors = all signaling mediators
* Save outputs and objects
  
# Load packages
```{r}
# Time tracking
ptm <- proc.time()

# Setting seed
set.seed(42)

# Load packages
suppressPackageStartupMessages({
  library(multinichenetr)
  library(nichenetr)
  library(igraph)
  library(tidyverse)
  library(readr)
  library(gprofiler2)
  library(here)
  library(styler)
  library(lintr)
})

source(here("src", "functions_CCCin3xTgAD.R"))
```

# Load data
```{r}
# MultiNicheNet output
output <- readRDS(here("data", "ccc", "multinichenet_output.rds"))
```

# Load & prepare AD gene lists
* Information on the gene lists can be found in doc/README
```{r}
# HP
## load, grab "GENE_SYMBOLS" row, split by comma, and remove empty strings
hp <- read_tsv(here("doc", "HP_ALZHEIMER_DISEASE.v2023.2.Hs.tsv")) %>%
  column_to_rownames("STANDARD_NAME")

hp <- hp["GENE_SYMBOLS", ]

hp <- unlist(strsplit(hp, ","))

hp <- hp[hp != ""]

# KEGG
## load, grab "GENE_SYMBOLS" row, split by comma, and remove empty strings
kegg <- read_tsv(here("doc", "KEGG_ALZHEIMERS_DISEASE.v2023.2.Hs.tsv")) %>%
  column_to_rownames("STANDARD_NAME")

kegg <- kegg["GENE_SYMBOLS", ]

kegg <- unlist(strsplit(kegg, ","))

kegg <- kegg[kegg != ""]

# GWAS
gwas <- read_csv(here("doc", "ad_gene_list.csv")) %>%
  pull(Gene)

# combine into 1 gene list
ad_genes <- c(hp, kegg, gwas) %>%
  unique()
```

# Make & Save Table S4
```{r}
kegg <- kegg %>%
  as.data.frame() %>%
  rename("gene" = ".") %>%
  mutate(origin = "msigdb: KEGG_ALZHEIMERS_DISEASE")

hp <- hp %>%
  as.data.frame() %>%
  rename("gene" = ".") %>%
  mutate(origin = "msigdb: HP_ALZHEIMER_DISEASE")

gwas <- gwas %>%
  as.data.frame() %>%
  rename("gene" = ".") %>%
  mutate(origin = "GWAS: Bellenguez, et al. (2022)")

table_s4 <- rbind(
  kegg,
  hp,
  gwas
)

WriteXLS::WriteXLS(table_s4, here(
  "results",
  "final_outputs",
  "01_figures",
  "table_s4.xlsx"
))
```

# Convert AD gene list to mouse & save
* Gene conversion did produce reproducible results, therefore, please load in the ad gene list we previously generated
    * Generated and converted on: 240210
```{r}
ad_genes_mm <- gorth(ad_genes,
  source_organism = "hsapiens",
  target_organism = "mmusculus"
)$ortholog_name

# saving both converted gene sets for later use
saveRDS(ad_genes_mm, file = here("doc", "ad_genes_mm.rds"))
```

# Load previously generated AD gene list
```{r}
ad_genes_mm <- readRDS(here("doc", "ad_genes_mm.rds"))
```

# NicheNet-v2 prior
* downloaded here: https://zenodo.org/record/7074291
* Accessed on 231211
```{r}
lr_network <- readRDS(here(
  "data",
  "ccc",
  "nichenet_v2_prior",
  "lr_network_mouse_21122021.rds"
)) %>%
  dplyr::rename(ligand = from, receptor = to) %>%
  distinct(ligand, receptor) %>%
  mutate(ligand = make.names(ligand), receptor = make.names(receptor))

ligand_target_matrix <- readRDS(here(
  "data",
  "ccc",
  "nichenet_v2_prior",
  "ligand_target_matrix_nsga2r_final_mouse.rds"
))

# format row and column names
colnames(ligand_target_matrix) <- colnames(ligand_target_matrix) %>%
  make.names()
rownames(ligand_target_matrix) <- rownames(ligand_target_matrix) %>%
  make.names()
```

# MultiNicheNet GRN prior
* All of the priors were downloaded from the MultiNicheNet zenodo
    * Accesssed 231211
* As described in the vignette, we convert the names of the genes inside the prior using make.names
```{r}
sig_network <- readRDS(
  here(
    "data",
    "ccc",
    "nichenet_v2_prior",
    "signaling_network_mouse_21122021.rds"
  )
) %>%
  mutate(
    from = make.names(from),
    to = make.names(to)
  )

gr_network <- readRDS(
  here(
    "data",
    "ccc",
    "nichenet_v2_prior",
    "gr_network_mouse_21122021.rds"
  )
) %>%
  mutate(
    from = make.names(from),
    to = make.names(to)
  )

ligand_tf_matrix <- readRDS(
  here(
    "data",
    "ccc",
    "nichenet_v2_prior",
    "ligand_tf_matrix_nsga2r_final_mouse.rds"
  )
)

colnames(ligand_tf_matrix) <- colnames(ligand_tf_matrix) %>%
  make.names()
rownames(ligand_tf_matrix) <- rownames(ligand_tf_matrix) %>%
  make.names()

weighted_networks <- readRDS(
  here(
    "data",
    "ccc",
    "nichenet_v2_prior",
    "weighted_networks_nsga2r_final_mouse.rds"
  )
)

weighted_networks$lr_sig <- weighted_networks$lr_sig %>%
  mutate(
    from = make.names(from),
    to = make.names(to)
  )
weighted_networks$gr <- weighted_networks$gr %>%
  mutate(
    from = make.names(from),
    to = make.names(to)
  )
```

# Set MultiNicheNet filtering parameters
```{r}
# number of targets
top_n_target <- 250

# contrast table
contrast_tbl <- tibble(
  contrast = c("X6mAD-X6mWT", "X6mWT-X6mAD", "X12mAD-X12mWT", "X12mWT-X12mAD"),
  group = c("X6mAD", "X6mWT", "X12mAD", "X12mWT")
)

# set receiver and sender cell types of interest
receiver_oi <- c("Excitatory.Neurons", "Inhibitory.Neurons")
sender_oi <- c("Astrocytes", "Microglia", "OPCs", "Oligodendrocytes")

# set groups to filter for (in this case all of them)
group_oi <- c("X6mWT", "X6mAD", "X12mWT", "X12mAD")
```

# Filter MultiNicheNet outputs
```{r}
output_filt <- filter_nichenet(output) %>%
  mutate(time_point = ifelse(grepl("X6m", group), "6m", "12m"))
```

# Get AD-associated LRTs
```{r}
ccc_filt <- output_filt %>%
  filter(target %in% ad_genes_mm) %>%
  select(sender, receiver, ligand, receptor, target, id_target, time_point) %>%
  unique()
```

# Split by time point
```{r}
objects_filt <- split(ccc_filt, ccc_filt$time_point)
```

# Make igraph objects
* I am also plotting all the sub-networks and saving to specified directory for each dataset
```{r}
igraph_objects_list <- signaling_igraph(
  filtered_df = objects_filt,
  ligand_tf_matrix = ligand_tf_matrix,
  weighted_networks = weighted_networks,
  lr_network = lr_network,
  sig_network = sig_network,
  gr_network = gr_network,
  plots = "results/intermediate_outputs/05_grn/"
)

# save list of igraph objects
saveRDS(
  igraph_objects_list,
  here(
    "results",
    "intermediate_outputs",
    "05_grn",
    "igraph_objects_list.rds"
  )
)

# unlist objects
list2env(igraph_objects_list, globalenv())
```

# Calculate network properties
* I calculated many network topology metrics, however the one highlighted in the manuscript is:
  * neighbors: all nodes directly connected to our receptor node. Must be outgoing from the receptor.
  
* NOTE: This chunk produces warnings due to missing information to calculate shortest path. That is expected.
```{r include=FALSE}
network_proporties_6m <- network_topology(igraph_objects_6m)

network_proporties_12m <- network_topology(igraph_objects_12m)
```

# Save outputs
```{r}
saveRDS(network_proporties_6m, file = here(
  "results",
  "intermediate_outputs",
  "05_grn",
  "signaling_grn_properties_6m.rds"
))

saveRDS(network_proporties_12m, file = here(
  "results",
  "intermediate_outputs",
  "05_grn",
  "signaling_grn_properties_12m.rds"
))
```

# Save igraph objects
```{r}
saveRDS(igraph_objects_6m, file = here(
  "data",
  "ccc",
  "6m_signaling_igraph_objects.rds"
))

saveRDS(igraph_objects_12m, file = here(
  "data",
  "ccc",
  "12m_signaling_igraph_objects.rds"
))
```

# Session info
```{r}
sessionInfo() # see output below
```
R version 4.2.3 (2023-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lintr_3.0.2          styler_1.9.1         here_1.0.1           lubridate_1.9.2     
 [5] forcats_1.0.0        stringr_1.5.0        dplyr_1.1.1          purrr_1.0.1         
 [9] readr_2.1.4          tidyr_1.3.0          tibble_3.2.1         ggplot2_3.4.2       
[13] tidyverse_2.0.0      igraph_1.4.2         nichenetr_2.0.4      multinichenetr_1.0.3

loaded via a namespace (and not attached):
  [1] rsvd_1.0.5                  Hmisc_5.0-1                 ica_1.0-3                  
  [4] ps_1.7.5                    class_7.3-21                rprojroot_2.0.3            
  [7] foreach_1.5.2               lmtest_0.9-40               crayon_1.5.2               
 [10] rbibutils_2.2.13            MASS_7.3-58.3               nlme_3.1-162               
 [13] backports_1.4.1             sva_3.46.0                  rlang_1.1.0                
 [16] XVector_0.38.0              caret_6.0-94                ROCR_1.0-11                
 [19] irlba_2.3.5.1               callr_3.7.3                 nloptr_2.0.3               
 [22] limma_3.54.2                scater_1.26.1               BiocParallel_1.32.6        
 [25] rjson_0.2.21                bit64_4.0.5                 glue_1.6.2                 
 [28] sctransform_0.4.1           processx_3.8.1              pbkrtest_0.5.2             
 [31] parallel_4.2.3              vipor_0.4.5                 spatstat.sparse_3.0-1      
 [34] AnnotationDbi_1.60.2        BiocGenerics_0.44.0         UpSetR_1.4.0               
 [37] muscat_1.12.1               spatstat.geom_3.1-0         tidyselect_1.2.0           
 [40] SummarizedExperiment_1.28.0 SeuratObject_4.1.3          fitdistrplus_1.1-8         
 [43] variancePartition_1.28.9    XML_3.99-0.14               zoo_1.8-12                 
 [46] ggpubr_0.6.0                xtable_1.8-4                magrittr_2.0.3             
 [49] evaluate_0.20               Rdpack_2.4                  scuttle_1.8.4              
 [52] cli_3.6.1                   zlibbioc_1.44.0             rstudioapi_0.14            
 [55] miniUI_0.1.1.1              sp_1.6-0                    rpart_4.1.19               
 [58] aod_1.3.2                   locfdr_1.1-8                shiny_1.7.4                
 [61] BiocSingular_1.14.0         xfun_0.38                   clue_0.3-64                
 [64] cluster_2.1.4               caTools_1.18.2              tidygraph_1.2.3            
 [67] KEGGREST_1.38.0             clusterGeneration_1.3.7     ggrepel_0.9.3              
 [70] listenv_0.9.0               Biostrings_2.66.0           png_0.1-8                  
 [73] future_1.32.0               ipred_0.9-14                withr_2.5.0                
 [76] bitops_1.0-7                ggforce_0.4.1               plyr_1.8.8                 
 [79] hardhat_1.3.0               dqrng_0.3.0                 e1071_1.7-13               
 [82] pROC_1.18.0                 coda_0.19-4                 pillar_1.9.0               
 [85] gplots_3.1.3                GlobalOptions_0.1.2         cachem_1.0.7               
 [88] GetoptLong_1.0.5            RUnit_0.4.32                DelayedMatrixStats_1.20.0  
 [91] vctrs_0.6.2                 ellipsis_0.3.2              generics_0.1.3             
 [94] lava_1.7.2.1                tools_4.2.3                 foreign_0.8-84             
 [97] remaCor_0.0.11              beeswarm_0.4.0              munsell_0.5.0              
[100] tweenr_2.0.2                emmeans_1.8.5               proxy_0.4-27               
[103] DelayedArray_0.24.0         fastmap_1.1.1               compiler_4.2.3             
[106] abind_1.4-5                 httpuv_1.6.9                plotly_4.10.1              
[109] GenomeInfoDbData_1.2.9      prodlim_2023.03.31          gridExtra_2.3              
[112] glmmTMB_1.1.7               edgeR_3.40.2                lattice_0.21-8             
[115] ggnewscale_0.4.8            deldir_1.0-6                visNetwork_2.1.2           
[118] utf8_1.2.3                  later_1.3.0                 recipes_1.0.5              
[121] jsonlite_1.8.4              scales_1.2.1                ScaledMatrix_1.6.0         
[124] carData_3.0-5               pbapply_1.7-0               sparseMatrixStats_1.10.0   
[127] estimability_1.4.1          genefilter_1.80.3           lazyeval_0.2.2             
[130] promises_1.2.0.1            car_3.1-2                   doParallel_1.0.17          
[133] R.utils_2.12.2              goftest_1.2-3               spatstat.utils_3.0-2       
[136] reticulate_1.28             checkmate_2.1.0             cyclocomp_1.1.0            
[139] rmarkdown_2.21              cowplot_1.1.1               blme_1.0-5                 
[142] statmod_1.5.0               Rtsne_0.16                  Biobase_2.58.0             
[145] uwot_0.1.14                 yaml_2.3.7                  survival_3.5-5             
[148] numDeriv_2016.8-1.1         htmltools_0.5.5             memoise_2.0.1              
[151] Seurat_4.3.0                locfit_1.5-9.7              graphlayouts_0.8.4         
[154] IRanges_2.32.0              viridisLite_0.4.1           digest_0.6.31              
[157] RhpcBLASctl_0.23-42         mime_0.12                   RSQLite_2.3.1              
[160] future.apply_1.10.0         remotes_2.4.2               data.table_1.14.8          
[163] blob_1.2.4                  R.oo_1.25.0                 S4Vectors_0.36.2           
[166] DiagrammeR_1.0.9            labeling_0.4.2              splines_4.2.3              
[169] Formula_1.2-5               RCurl_1.98-1.12             broom_1.0.4                
[172] hms_1.1.3                   colorspace_2.1-0            base64enc_0.1-3            
[175] ggbeeswarm_0.7.1            GenomicRanges_1.50.2        shape_1.4.6                
[178] nnet_7.3-18                 Rcpp_1.0.10                 RANN_2.6.1                 
[181] mvtnorm_1.1-3               circlize_0.4.15             fansi_1.0.4                
[184] tzdb_0.3.0                  parallelly_1.35.0           ModelMetrics_1.2.2.2       
[187] R6_2.5.1                    grid_4.2.3                  factoextra_1.0.7           
[190] ggridges_0.5.4              lifecycle_1.0.3             bluster_1.8.0              
[193] ggsignif_0.6.4              minqa_1.2.5                 leiden_0.4.3               
[196] Matrix_1.5-4                desc_1.4.2                  RcppAnnoy_0.0.20           
[199] RColorBrewer_1.1-3          iterators_1.0.14            spatstat.explore_3.1-0     
[202] TMB_1.9.4                   gower_1.0.1                 R.cache_0.16.0             
[205] htmlwidgets_1.6.2           beachmat_2.14.2             polyclip_1.10-4            
[208] shadowtext_0.1.2            timechange_0.2.0            mgcv_1.8-42                
[211] ComplexHeatmap_2.14.0       globals_0.16.2              htmlTable_2.4.1            
[214] patchwork_1.1.2             spatstat.random_3.1-4       progressr_0.13.0           
[217] codetools_0.2-19            matrixStats_0.63.0          metapod_1.6.0              
[220] randomForest_4.7-1.1        gtools_3.9.4                prettyunits_1.1.1          
[223] SingleCellExperiment_1.20.1 R.methodsS3_1.8.2           GenomeInfoDb_1.34.9        
[226] gtable_0.3.3                DBI_1.1.3                   stats4_4.2.3               
[229] tensor_1.5                  httr_1.4.5                  KernSmooth_2.23-20         
[232] stringi_1.7.12              progress_1.2.2              reshape2_1.4.4             
[235] farver_2.1.1                annotate_1.76.0             viridis_0.6.2              
[238] rex_1.2.1                   fdrtool_1.2.17              timeDate_4022.108          
[241] xml2_1.3.3                  boot_1.3-28.1               BiocNeighbors_1.16.0       
[244] lme4_1.1-32                 geneplotter_1.76.0          scattermore_0.8            
[247] scran_1.26.2                DESeq2_1.38.3               bit_4.0.5                  
[250] MatrixGenerics_1.10.0       spatstat.data_3.0-1         ggraph_2.1.0               
[253] pkgconfig_2.0.3             lmerTest_3.1-3              rstatix_0.7.2              
[256] knitr_1.42

# Time tracking
```{r}
fptm <- proc.time() - ptm

# convert to minutes
(fptm[3] / 60)

# Time elapsed: 15 minutes
```

# Reproducibility
```{r}
# styler
style_file(here(
  "src",
  "ccc",
  "04_grn.Rmd"
))

# lintr
lint(filename = here(
  "src",
  "ccc",
  "04_grn.Rmd"
))
```
